traits/trait_list_object.py

```diff
@@ -53,460 +94,754 @@ def __init__(self, index=0, removed=None, added=None):
+            removed : list
+                Will be []
+
+        """
+        iterable = list(iterable)
+        index = slice(len(self), len(self) + len(iterable))
+        removed = []
+        added = self.validate(index, removed, iterable)
+
+        super().extend(added)
+
+        if self._should_notify(removed, added):
+            self.notify(index, removed, added)
+
+    def insert(self, index, object):
+        """ Insert an object to the list before index.
+
+        Parameters
+        ----------
+        index : int
+            The index before which to insert the object.
+        object : Any
+            The object to insert.
+
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The index before which the item was inserted.
+            added : list
+                The newly added item(s) as a list.
+            removed : list
+                Will be []
+
+        """
+        removed = []
+        norm_index = self._normalize_index(index)
+        added = self.validate(norm_index, removed, [object])
 
-        if trait.minlen <= (len(self) + len_xlist) <= trait.maxlen:
-            object = self.object()
-            name = self.name
-            validate = trait.item_trait.handler.validate
-            try:
-                if validate is not None:
-                    xlist = [validate(object, name, value) for value in xlist]
+        super().insert(index, added[0])
 
-                list.extend(self, xlist)
+        if self._should_notify(removed, added):
+            self.notify(norm_index, removed, added)
+
+    def clear(self):
+        """ Remove all items from the list
 
-                if (self.name_items is not None) and (len(xlist) != 0):
-                    self._send_trait_items_event(
-                        self.name_items,
-                        TraitListEvent(len(self) - len(xlist), None, xlist),
-                        trait.items_event(),
-                    )
 
-                return
+        Notes
+        -----
+        Notification:
+            index : slice
+                The range of the slice will be 0 to len(self).
+            added : list
+                The empty list [].
+            removed : list
+                The removed items.
+
+        """
+        index = slice(0, len(self), None)
+        removed = self.copy()
+        self.validate(index, removed, [])
+        super().clear()
+
+        if self._should_notify(removed, []):
+            self.notify(index, removed, [])
+
+    def pop(self, index=-1):
+        """ Remove and return item at index (default last).
+
+        Parameters
+        ----------
+        index: int
+            Index at which item has to be removed.
+
+        Returns
+        -------
+        item : An item in the list at given index, last item if no index given.
+
+        Raises
+        ------
+        IndexError
+            If list is empty or index is out of range.
```

Comment:
Nit: I think the "list is empty or" part is redundant and can be dropped: that's just a special case of the index being out of range.